### ISSUE TITLE
Pylance generic type error

### ISSUE THREAD

**Issue by PASUNX** [microsoft/pylance-release#1121]
<h3> Environment data </h3>

-   Language Server version: 2021.3.4
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: strict

<h3> Code Snippet / Additional information </h3>

```python
class A:

    class B: ...

    a = list["A.B"]()   # ok but no color
    b: list["A.B"] = [] # A is class but B is Unknown

A.a.append(A.B()) # ok
A.b.append(A.B()) # error:
                  #    Argument of type "B" cannot be assigned to parameter "__object" of type "Literal['A.B']" in function "append"
                  #      "B" cannot be assigned to type "Literal['A.B']" Pylance (reportGeneralTypeIssues)
```
![image](https://user-images.githubusercontent.com/76944694/113502094-283fe680-9554-11eb-8b64-730bd5460c95.png)

**Comment by erictraut** [comment]
Thanks for reporting the problem.

This appears to be a problem in the order of expression evaluation within pyright. The semantic highlighting is causing a different evaluation order than would otherwise occur, and this exposes a bug in the type evaluator. The problem doesn't occur (or is much harder to trigger) when semantic highlighting is disabled. I'll investigate further.

**Comment by erictraut** [comment]
I found the source of the bug. It was a regression that I introduced about two months ago when adding support for [PEP 637](https://www.python.org/dev/peps/pep-0637/). It resulted in inconsistent type evaluations of string literals within type annotations based on the order in which nodes were evaluated. Pylance's semantic highlighting feature caused nodes to be visited in a different order than they otherwise would be, and that exposed the bug.

This will be fixed in the next release of Pylance.
